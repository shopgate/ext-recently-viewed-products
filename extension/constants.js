module.exports = {
  STORAGE_RECENTLY_VIEWED_PRODUCTS_LIST: 'recently-viewed-products-list'
}
