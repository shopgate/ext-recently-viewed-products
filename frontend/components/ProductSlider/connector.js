import { connect } from 'react-redux';
import { getPageUrl } from '../../selectors';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  showMoreUrl: getPageUrl(state),
});

export default connect(mapStateToProps);
