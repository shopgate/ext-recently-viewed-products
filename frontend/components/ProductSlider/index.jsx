import React, { Component } from 'react';
import PropTypes from 'prop-types';
import I18n from '@shopgate/pwa-common/components/I18n';
import Slider from '@shopgate/pwa-common/components/Slider';
import ButtonLink from '@shopgate/pwa-ui-shared/ButtonLink';
import { showOnPdpPage, showOnEmptyCartPage } from '../../config';
import Card from '../Card';
import ProductCard from '../ProductCard';
import connect from './connector';
import styles from './style';

/**
 * Creates an item for a single product.
 * @param {Object} product The product data.
 * @return {JSX} The rendered product card.
 */
const createSliderItem = (product) => {
  const key = `s${product.id}`;
  return (
    <Slider.Item key={key} className={styles.sliderItem}>
      <Card className={styles.card}>
        <ProductCard
          product={product}
          titleRows={2}
          hidePrice
          hideRating
        />
      </Card>
    </Slider.Item>
  );
};

/**
 * The ProductSlider component
 */
class ProductSlider extends Component {
  static propTypes = {
    autoPlay: PropTypes.bool,
    headline: PropTypes.string,
    isCartPage: PropTypes.bool,
    isProductPage: PropTypes.bool,
    products: PropTypes.arrayOf(PropTypes.shape()),
    showMore: PropTypes.bool,
    showMoreUrl: PropTypes.string,
  };

  static defaultProps = {
    autoPlay: false,
    products: [],
    showMore: false,
    showMoreUrl: null,
    isCartPage: false,
    isProductPage: false,
    headline: null,
  };

  /**
   * Should component update given the new props?
   * @param {Object} nextProps The next component props.
   * @return {boolean} Update or not.
   */
  shouldComponentUpdate(nextProps) {
    if (nextProps.isCartPage !== this.props.isCartPage) {
      return true;
    }

    if (nextProps.isProductPage !== this.props.isProductPage) {
      return true;
    }

    if (nextProps.products.length !== this.props.products.length) {
      return true;
    }

    return this.props.products.every(({ id }, index) => id !== nextProps.products[index].id);
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    if (this.props.isCartPage && !showOnEmptyCartPage) {
      return null;
    }

    if (this.props.isProductPage && !showOnPdpPage) {
      return null;
    }

    const items = this.props.products.map((
      product => createSliderItem(product)
    ));

    if (!items.length) {
      return null;
    }

    let headline = 'recently_viewed_products.headline';

    if (this.props.isCartPage) {
      headline += '_cart';
    }

    const hasShowMore = this.props.showMore && this.props.showMoreUrl;
    return (
      <div className={styles.slider}>
        <div className={styles.headlineContainer}>
          <h3 className={styles.headline(hasShowMore)}>
            <I18n.Text string={this.props.headline ? this.props.headline : headline} />
          </h3>
          { hasShowMore && (
            <div className={styles.showMoreContainer}>
              <ButtonLink href={this.props.showMoreUrl} noGap>
                <I18n.Text string="recently_viewed_products.show_more" />
              </ButtonLink>
            </div>
          )}
        </div>

        <Slider
          autoPlay={this.props.autoPlay}
          loop={false}
          indicators={false}
          controls={false}
          snapItems={false}
          slidesPerView={2.3}
          classNames={{ container: styles.sliderContainer }}
        >
          {items}
        </Slider>
      </div>
    );
  }
}

export default connect(ProductSlider);
