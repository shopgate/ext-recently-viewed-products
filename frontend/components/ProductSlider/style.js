import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';
import { isIOSTheme } from '@shopgate/pwa-extension-kit/env/helpers';

const { colors, variables } = themeConfig;

const sliderContainer = css({
  marginLeft: 'auto',
  marginRight: 'auto',
  position: 'relative',
  paddingBottom: variables.gap.big,
}).toString();

const slider = css({
  width: '100%',
  flex: 1,
  padding: `${variables.gap.small}px 0`,
}).toString();

const sliderItem = css({
  width: '50%',
}).toString();

const productInfo = css({
  padding: 15,
}).toString();

const productName = css({
  fontWeight: 500,
}).toString();

const priceGrid = css({
  alignItems: 'flex-end',
}).toString();

const priceStrikedItem = css({
  flexGrow: 1,
}).toString();

const priceItem = css({
  flexGrow: 1,
  textAlign: 'right',
}).toString();

const priceBase = css({
  padding: '0 15px',
}).toString();

const priceStriked = css({
  ...priceBase,
  fontSize: '0.875rem',
}).toString();

const price = css({
  ...priceBase,
  color: colors.primary,
  fontSize: '1rem',
}).toString();

const card = css({
  background: '#FFF',
  height: '100%',
  margin: '0px 8px',
}).toString();

const headlineContainer = css({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  padding: '8px 0',
  overflow: 'hidden',
});

/**
 * Headline css
 * @param {boolean} hasShowMore Whether there is more button visible.
 * @returns {string}
 */
const headline = hasShowMore => css({
  fontSize: isIOSTheme() ? 22 : 18,
  width: '100%',
  // The Button which is right aligned to the headline has a paddingTop modification by 1px
  margin: isIOSTheme() ? '8px 16px' : '8px 8px 8px 12px',
  padding: 0,
  textAlign: hasShowMore || isIOSTheme() ? 'left' : 'center',
}).toString();

const showMoreContainer = css({
  ' button': {
    fontSize: 'medium',
  },
});

export default {
  card,
  headlineContainer,
  headline,
  sliderContainer,
  slider,
  sliderItem,
  productInfo,
  productName,
  priceGrid,
  priceStrikedItem,
  priceItem,
  priceBase,
  priceStriked,
  price,
  showMoreContainer,
};
