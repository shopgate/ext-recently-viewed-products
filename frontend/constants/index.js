export const PIPELINE_ADD_VIEWED_PRODUCTS = 'shopgate.user.addViewedProducts';
export const PIPELINE_GET_VIEWED_PRODUCTS = 'shopgate.user.getViewedProducts';

export const GET_VIEWED_PRODUCTS_LIMIT = 50;
export const RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT = 4;

export const REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS = '@shopgate/recently-viewed-products/productList';

export const REQUEST_RECENTLY_VIEWED_PRODUCTS = 'REQUEST_RECENTLY_VIEWED_PRODUCTS';
export const RESPONSE_RECENTLY_VIEWED_PRODUCTS = 'RESPONSE_RECENTLY_VIEWED_PRODUCTS';
export const ERROR_RECENTLY_VIEWED_PRODUCTS = 'ERROR_RECENTLY_VIEWED_PRODUCTS';

export const RECENTLY_VIEWED_PATH = '/recently-viewed';
