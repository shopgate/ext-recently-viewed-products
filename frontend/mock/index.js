import configureStore from 'redux-mock-store';
import { QUICKLINKS_MENU } from '@shopgate/pwa-common/constants/MenuIDs';
import { REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS } from '../constants';
import { pageId } from '../config';

export const defaultState = {
  extensions: {
    [REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS]: {
      isFetching: false,
      productIds: [],
    },
  },
  product: {
    currentProduct: { productId: null },
  },
  history: { pathname: '' },
  menu: {
    menusById: {

    },
  },
};

/**
 * Creates a mocked store with the passed amount of products.
 * @param {number} [amount=2] The amount of products.
 * @param {boolean} [createPageEntry=true] Tells if an entry for a recently viewed products page
 *   shall be added.
 * @return {Object}
 */
export const createStateWithProducts = (amount = 2, createPageEntry = true) => {
  const productIds = [];
  const productsById = {};

  for (let i = 1; i <= amount; i += 1) {
    productIds.push(i.toString());
    productsById[i] = {
      productData: {
        id: i.toString(),
        featuredImageUrl: `https://example.com/${i}`,
        name: `${i}`,
      },
    };
  }

  return {
    product: {
      productsById,
    },
    extensions: {
      [REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS]: {
        isFetching: false,
        productIds,
      },
    },
    menu: {
      menusById: {
        [QUICKLINKS_MENU]: {
          isFetching: false,
          entries: [{
            url: `/page/${createPageEntry ? pageId : 'noop'}`,
            label: 'Recently viewed',
          }],
        },
      },
    },
  };
};

export const stateWithoutValidMenuEntries = {
  menu: {
    menusById: {
      [QUICKLINKS_MENU]: {
        isFetching: false,
        entries: [{
          url: '/page/foobar',
          label: 'Recently viewed',
        }],
      },
    },
  },
};

/**
 * Empty store.
 * @returns {function}
 */
export const getEmptyStore = () => configureStore()(defaultState);

// eslint-disable-next-line require-jsdoc
export const getStoreWithProducts = (amount = 2, createPageEntry = true) =>
  configureStore()(createStateWithProducts(amount, createPageEntry));
