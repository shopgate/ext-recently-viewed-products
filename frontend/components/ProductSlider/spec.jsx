import React, { Component as mockedComponent } from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { getEmptyStore, getStoreWithProducts } from '../../mock';
import { RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT } from '../../constants';
import ProductSlider from './index';

/* eslint-disable require-jsdoc, global-require */
jest.mock('@shopgate/pwa-common/components/Slider', () => (class extends mockedComponent {
  static Item({ children }) { return (<div>{children}</div>); }
  render() {
    return (<div>{this.props.children}</div>);
  }
}));

jest.mock('@shopgate/pwa-common-commerce/market/helpers/showTaxDisclaimer', () => () => null);

jest.mock(
  '@shopgate/pwa-common/helpers/config',
  () => require('@shopgate/pwa-common/helpers/config/mock')
);
/* eslint-enable */

describe('ProductSlider', () => {
  it('should render nothing', () => {
    const component = mount((
      <Provider store={getEmptyStore()}>
        <ProductSlider />
      </Provider>
    ));

    expect(component).toMatchSnapshot();
    expect(component.html()).toBe(null);
  });

  it('should render slider', () => {
    const component = mount((
      <Provider store={getStoreWithProducts()}>
        <ProductSlider />
      </Provider>
    ));

    expect(component).toMatchSnapshot();
    expect(component.find('Item').length).toBe(2);
    expect(component.find('ButtonLink').length).toBe(0);
  });

  it('should render slider with a show more button', () => {
    const component = mount((
      <Provider store={getStoreWithProducts(RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT + 1)}>
        <ProductSlider />
      </Provider>
    ));

    expect(component).toMatchSnapshot();
    expect(component.find('Item').length).toBe(RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT);
    expect(component.find('ButtonLink').length).toBe(1);
  });
});
