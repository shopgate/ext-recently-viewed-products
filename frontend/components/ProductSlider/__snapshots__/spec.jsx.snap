// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSlider should render nothing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductSlider)
    products={Array []}
  >
    <ProductSlider
      autoPlay={false}
      dispatch={[Function]}
      headline={null}
      isCartPage={false}
      isProductPage={false}
      products={Array []}
      showMore={false}
      showMoreUrl={null}
    />
  </Connect(ProductSlider)>
</Provider>
`;

exports[`ProductSlider should render slider 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductSlider)
    isCartPage={true}
    products={
      Array [
        Object {
          "featuredImageUrl": "https://example.com/1",
          "id": "1",
          "name": "1",
        },
        Object {
          "featuredImageUrl": "https://example.com/2",
          "id": "2",
          "name": "2",
        },
      ]
    }
    showMore={false}
  >
    <ProductSlider
      autoPlay={false}
      dispatch={[Function]}
      headline={null}
      isCartPage={true}
      isProductPage={false}
      products={
        Array [
          Object {
            "featuredImageUrl": "https://example.com/1",
            "id": "1",
            "name": "1",
          },
          Object {
            "featuredImageUrl": "https://example.com/2",
            "id": "2",
            "name": "2",
          },
        ]
      }
      showMore={false}
      showMoreUrl="/page/recently-viewed"
    />
  </Connect(ProductSlider)>
</Provider>
`;

exports[`ProductSlider should render slider with a show more button 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductSlider)
    isCartPage={true}
    products={
      Array [
        Object {
          "featuredImageUrl": "https://example.com/1",
          "id": "1",
          "name": "1",
        },
        Object {
          "featuredImageUrl": "https://example.com/2",
          "id": "2",
          "name": "2",
        },
      ]
    }
    showMore={true}
  >
    <ProductSlider
      autoPlay={false}
      dispatch={[Function]}
      headline={null}
      isCartPage={true}
      isProductPage={false}
      products={
        Array [
          Object {
            "featuredImageUrl": "https://example.com/1",
            "id": "1",
            "name": "1",
          },
          Object {
            "featuredImageUrl": "https://example.com/2",
            "id": "2",
            "name": "2",
          },
        ]
      }
      showMore={true}
      showMoreUrl="/page/recently-viewed"
    />
  </Connect(ProductSlider)>
</Provider>
`;
